### DISCUSSION TITLE
How to add u8g2 to keil u-vision project with LPC1768 microcontroller

### DISCUSSION THREAD

**Discussion by azizolahi** [olikraus/u8g2#2142] [Q&A]
Hi dear,
I am beginner at LPC1768 and jlx256128g please guide me how to use u8g2 c library step by step.
Thanks a lot.

**Comment by olikraus** [comment]
This project is mostly for Arduino Environment. However there is a pure C layer, which can be used to port to other systems.
The biggest issue will be the access to your low level COM hardware (I2C/SPI). This is described here: https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform
So I suggest to read though this page first.

There is also an example for a simple bit-banging I2C for the LPC here: https://github.com/olikraus/u8g2/blob/master/sys/arm/lpc824/u8g2_logo/u8x8cb.c

**Comment by pillopaolo** [comment]
Hi Oliver,

I finally managed to implement your super u8g2 with:
- LPC1768 controller - SKR 1.4 board
- ST7920 display - standard Creality display
-  3 wires

However, I had the following issue:
- I started with using `u8x8_byte_3wire_sw_spi` in the following code, WITHOUT success:
  `u8g2_Setup_st7920_s_128x64_f(&u8g2, rotation, u8x8_byte_3wire_sw_spi, u8x8_gpio_and_delay_lpc1768);`
  `u8x8_SetPin_3Wire_SW_SPI(getU8x8(), clock, data, cs, reset)`
   where:
   `u8x8_byte_3wire_sw_spi` is the standard bit-banged interface found in u8x8_byte.c
   `u8x8_gpio_and_delay_lpc1768` is my custom GPIO and Delay callback
- Then, out of desperation, I changed to `u8x8_byte_4wire_sw_spi`, and it magically WORKED!
  BTW, I kept using `u8x8_SetPin_3Wire_SW_SPI(getU8x8(), clock, data, cs, reset)` in the second line!

So, why do I need to use `u8x8_byte_4wire_sw_spi` when I have 3 wires and use `u8x8_SetPin_3Wire_SW_SPI(...)`?

BTW, I also noticed that you did something similar in the standard Arduino setup, in u8g2lib.h:
`class U8G2_ST7920_128X64_F_SW_SPI`
    ` ...`
    ` u8g2_Setup_st7920_s_128x64_f(&u8g2, rotation, u8x8_byte_arduino_4wire_sw_spi, u8x8_gpio_and_delay_arduino);`
    ` u8x8_SetPin_3Wire_SW_SPI(getU8x8(), clock, data, cs, reset);`

Thanks!
Paolo

**Reply by olikraus** [reply]
The name 4wire vs 3wire is a little bit misleading. In the context of displays, 3wire SPI means, that the CD (A0) signal is encoded into the bit stream as additional 9th bit (plus the normal 8 data bits).

Although the is also no CD/A0 signal for the ST7920 the CD / A0 information is transfered within an additional byte (if i remember correctly). As a consequence you must use the normal 4wire callback.

**Reply by pillopaolo** [reply]
> ...the CD / A0 information is transfered within an additional byte (if i remember correctly). 

I guess you mean "without". If so pls edit the above answer for future use
Thanks so much!

**Reply by olikraus** [reply]
The command/data (CD, A0) information is sent as an extra byte. It will prefix the data stream. Not sure whether the term "within" is correct or not. The point is: Instead of a an extra wire for the CD information, there is an extra byte inserted into the data stream.
